### PR TITLE
8302814: Delete unused CountLoopEnd instruct with CmpX

### PR DESCRIPTION
**Pull request by sunny868**
CountLoopEnd only for T_int, therefore the following instructs in riscv.ad are useless and should be deleted.

CountedLoopEnd cmp (CmpU op1 op2)
CountedLoopEnd cmp (CmpP op1 op2)
CountedLoopEnd cmp (CmpN op1 op2)
CountedLoopEnd cmp (CmpF op1 op2)
CountedLoopEnd cmp (CmpD op1 op2)

and CountedLoopEnd with CmpU on x86*.ad, aarch64.ad ar useless also.

Please help review it.

Thanks.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8302814](https://bugs.openjdk.org/browse/JDK-8302814): Delete unused CountLoopEnd instruct with CmpX

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [fefb9122](https://git.openjdk.org/jdk/pull/12648/files/fefb9122e8d73612a98978898d2c025c9744fb27)
 * [Feilong Jiang](https://openjdk.org/census#fjiang) (@feilongjiang - Author) ⚠️ Review applies to [fefb9122](https://git.openjdk.org/jdk/pull/12648/files/fefb9122e8d73612a98978898d2c025c9744fb27)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12648/head:pull/12648` \
`$ git checkout pull/12648`

Update a local copy of the PR: \
`$ git checkout pull/12648` \
`$ git pull https://git.openjdk.org/jdk.git pull/12648/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12648`

View PR using the GUI difftool: \
`$ git pr show -t 12648`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12648.diff">https://git.openjdk.org/jdk/pull/12648.diff</a>

</details>

### CODE REVIEW

**Comment by sunny868** [comment]
The failure `linux-cross-compile / build (arm)` is not related to this patch.

**Comment by feilongjiang** [comment]
Hi, did you run any tests on these changes? I agree with removing CmpP/CmpN/CmpF/CmpD instructs. But for CmpU, I have checked aarch64.ad, and it provides the U version of CountedLoopEnd, I think we should do some investigating into this:
https://github.com/openjdk/jdk/blob/43c71ddf923d442499449948f4bf8a7c79249af0/src/hotspot/cpu/aarch64/aarch64.ad#L16526-L16541

And I also found `LongCountedLoopEndNode` definition in `loopnode.hpp` but not used yet:
https://github.com/openjdk/jdk/blob/43c71ddf923d442499449948f4bf8a7c79249af0/src/hotspot/share/opto/loopnode.hpp#L356-L371

Maybe we could keep the Long version?

**Comment by sunny868** [comment]
1. I only did tier1-3 and openjvm2008 tests on the LoongArch64 architecture(Linux platform) and confirmed that none of these types were used, including CmpU.
2. Do you sure the U version of CountedLoopEnd be used on AArch64? I guess it is unused also but I will also confirm further.
3. Yes, I will update the Patch for keep the Long version. I'm also evaluating the performance benefits of `LongCountedLoopEndNode` to decide whether to implement it.

**Comment by feilongjiang** [comment]
> Do you sure the U version of CountedLoopEnd be used on AArch64? 

I'm not sure about that. I just found that CmpU version was defined in `aarch64.ad` too.

> I guess it is unused also but I will also confirm further.

Yes, we should check if CmpU was safe to remove. By the way, looks x86 also provides CmpU version of CountedeLoopEnd:
https://github.com/openjdk/jdk/blob/180b94c73e9ad17d57650d4c985d4104289052a9/src/hotspot/cpu/x86/x86_64.ad#L13056-L13069

**Comment by sunny868** [comment]
For bytecodes if_icmpxx(used by java`s for loop)、ifeq/ifne/iflt/ifle/ifgt/ifge, if_icmpxx(used by java`s while loop), C2 compiler created CmpINode not CmpUNode, so I think that should doesn't exist CountedLoopNode with CmpUNode.
<pre><code class="cpp">
// src/hotspot/share/opto/parse2.cpp

2661   case Bytecodes::_ifeq: btest = BoolTest::eq; goto handle_ifxx;                
2662   case Bytecodes::_ifne: btest = BoolTest::ne; goto handle_ifxx;                
2663   case Bytecodes::_iflt: btest = BoolTest::lt; goto handle_ifxx;                
2664   case Bytecodes::_ifle: btest = BoolTest::le; goto handle_ifxx;                
2665   case Bytecodes::_ifgt: btest = BoolTest::gt; goto handle_ifxx;                
2666   case Bytecodes::_ifge: btest = BoolTest::ge; goto handle_ifxx;                
2667   handle_ifxx:                                                                  
2668     // If this is a backwards branch in the bytecodes, add Safepoint            
2669     maybe_add_safepoint(iter().get_dest());                                     
2670     a = _gvn.intcon(0);                                                         
2671     b = pop();                                                                  
2672     c = _gvn.transform( new CmpINode(b, a) );                                   
2673     do_if(btest, c);                                                            
2674     break;                                                                      
2675                                                                                 
2676   case Bytecodes::_if_icmpeq: btest = BoolTest::eq; goto handle_if_icmp;        
2677   case Bytecodes::_if_icmpne: btest = BoolTest::ne; goto handle_if_icmp;        
2678   case Bytecodes::_if_icmplt: btest = BoolTest::lt; goto handle_if_icmp;        
2679   case Bytecodes::_if_icmple: btest = BoolTest::le; goto handle_if_icmp;        
2680   case Bytecodes::_if_icmpgt: btest = BoolTest::gt; goto handle_if_icmp;        
2681   case Bytecodes::_if_icmpge: btest = BoolTest::ge; goto handle_if_icmp;        
2682   handle_if_icmp:                                                               
2683     // If this is a backwards branch in the bytecodes, add Safepoint            
2684     maybe_add_safepoint(iter().get_dest());                                     
2685     a = pop();                                                                  
2686     b = pop();                                                                  
2687     c = _gvn.transform( new CmpINode( b, a ) );  
</code></pre>

**Review by vnkozlov — COMMENTED** [review]
What about `jmpLoopEndUCF*` instructions?

**Review by vnkozlov — CHANGES_REQUESTED** [review]
Also I think it is not correct to remove `CmpU` case. There are IGVN transformations which convert `CmpI` to `CmpU`:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/subnode.cpp#L1632

May be that is why next test failed in your GHA testing:
compiler/vectorization/runner/LoopRangeStrideTest.java

**Comment by vnkozlov** [comment]
I am taking back my next comment because I see this test fails in other PRs:

>May be that is why next test failed in your GHA testing:
>compiler/vectorization/runner/LoopRangeStrideTest.java

**Comment by sunny868** [comment]
> Also I think it is not correct to remove `CmpU` case. There are IGVN transformations which convert `CmpI` to `CmpU`: https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/subnode.cpp#L1632

Maybe it only for range check. see:

Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code) 
V  [libjvm.so+0x1867734]  BoolNode::Ideal(PhaseGVN*, bool)+0xb84  (subnode.cpp:1631)
V  [libjvm.so+0x1567f65]  PhaseGVN::transform_no_reclaim(Node*)+0x25  (phaseX.cpp:840)
V  [libjvm.so+0x124b4f5]  LibraryCallKit::generate_limit_guard(Node*, Node*, Node*, RegionNode*)+0x145  (library_call.cpp:891)
V  [libjvm.so+0x122085a]  LibraryCallKit::generate_string_range_check(Node*, Node*, Node*, bool)+0xea  (library_call.cpp:912)
V  [libjvm.so+0x12340c3]  LibraryCallKit::inline_string_copy(bool)+0x203  (library_call.cpp:1403)
V  [libjvm.so+0x1248ee4]  LibraryIntrinsic::generate(JVMState*)+0x3d4  (library_call.cpp:116)

**Comment by sunny868** [comment]
> What about `jmpLoopEndUCF*` instructions?

Thanks @vnkozlov for the reminder, I think `jmpLoopEndUCF*` is unused also.

**Review by vnkozlov — COMMENTED** [review]
I am not convinced about CmpU but I submitted testing for v03 to see results.

**Review by vnkozlov — APPROVED** [review]
My tier1-4, xcomp and stress testing passed. I can't complain now :)
You need second review.

File: src/hotspot/cpu/riscv/riscv.ad
Comment (feilongjiang) [inline src/hotspot/cpu/riscv/riscv.ad:9297]: I think this instruction should keep too.

**Review by feilongjiang — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/riscv/riscv.ad
Comment (sunny868) [reply src/hotspot/cpu/riscv/riscv.ad:9297]: The current count loop iterations is1000(setting by `LoopStripMiningIter` ), and if it is too large, it will cause the GC pause too long, so I don't think there is any point in implementing the counted loop for Long type. However, I still respect your opinion and keep `cmpL loop` in risv.ad.

**Review by feilongjiang — APPROVED** [review]
New changes look good, thanks.

**Review by vnkozlov — APPROVED** [review]
v04 update is good

**Comment by sunny868** [comment]
Thanks @vnkozlov @feilongjiang for review.

**Comment by sunny868** [comment]
@vnkozlov Do you have any comments on this patch and could you please sponsor it for me?

**Comment by vnkozlov** [comment]
@sunny868 please merge master

And please ask for sponsorship sooner. I did not see you needed it.

**Comment by vnkozlov** [comment]
My testing (tier1-4, xcomp, stress) passed with latest version.

**Pull request closed without merging**
